tensorflow_io/ignite/kernels/ggfs/ggfs_client.cc

```diff
@@ -0,0 +1,381 @@
+
+  return Status::OK();
+}
+
+Status GGFSClient::MkDirs(const string &path, const bool only_if_not_exists) {
+  TF_RETURN_IF_ERROR(
+      SendCommonRequestHeader(kMkDirsMethodId, 6 + path.length()));
+
+  TF_RETURN_IF_ERROR(client_->WriteByte(kStringVal));
+  TF_RETURN_IF_ERROR(client_->WriteInt(path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteData(
+      reinterpret_cast<const uint8_t *>(path.c_str()), path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteByte(only_if_not_exists));
+
+  TF_RETURN_IF_ERROR(ReceiveCommonResponseHeader());
+
+  return Status::OK();
+}
+
+Status GGFSClient::ListFiles(const string &path,
+                             std::vector<string> *out_files) {
+  TF_RETURN_IF_ERROR(
+      SendCommonRequestHeader(kListFilesMethodId, 5 + path.length()));
+
+  TF_RETURN_IF_ERROR(client_->WriteByte(kStringVal));
+  TF_RETURN_IF_ERROR(client_->WriteInt(path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteData(
+      reinterpret_cast<const uint8_t *>(path.c_str()), path.length()));
+
+  TF_RETURN_IF_ERROR(ReceiveCommonResponseHeader());
+
+  int32_t length;
+  TF_RETURN_IF_ERROR(client_->ReadInt(&length));
+
+  while (length > 0) {
+    uint8_t type;
+    TF_RETURN_IF_ERROR(client_->ReadByte(&type));
+    if (type != kStringVal)
+      return errors::Unknown(
+          "Method GGFSClient::ListFiles expects strings in response");
+
+    int32_t str_length;
+    TF_RETURN_IF_ERROR(client_->ReadInt(&str_length));
+
+    uint8_t str[str_length];
+    TF_RETURN_IF_ERROR(client_->ReadData(str, str_length));
+
+    out_files->push_back(MakeRelative(
+        string(reinterpret_cast<char *>(str), str_length), path + "/"));
+
+    length--;
+  }
+
+  return Status::OK();
+}
+
+Status GGFSClient::SendCommonRequestHeader(const uint8_t method_id,
+                                           const int32_t length) {
+  EstablishConnection();
+  TF_RETURN_IF_ERROR(client_->WriteInt(32 + length));
+  TF_RETURN_IF_ERROR(client_->WriteShort(kCustomProcessorOpcode));
+  TF_RETURN_IF_ERROR(client_->WriteLong(0));
+  TF_RETURN_IF_ERROR(client_->WriteByte(kStringVal));
+  TF_RETURN_IF_ERROR(client_->WriteInt(16));
+  TF_RETURN_IF_ERROR(client_->WriteData(
+      reinterpret_cast<const uint8_t *>("ML_MODEL_STORAGE"), 16));
+  TF_RETURN_IF_ERROR(client_->WriteByte(method_id));
+
+  return Status::OK();
+}
+
+Status GGFSClient::ReceiveCommonResponseHeader() {
+  int32_t res_len;
+  TF_RETURN_IF_ERROR(client_->ReadInt(&res_len));
+
+  int64_t req_id;
+  TF_RETURN_IF_ERROR(client_->ReadLong(&req_id));
+
+  int32_t status;
+  TF_RETURN_IF_ERROR(client_->ReadInt(&status));
+
+  if (status != 0) {
+    uint8_t err_msg_header;
+    TF_RETURN_IF_ERROR(client_->ReadByte(&err_msg_header));
+    if (err_msg_header == kStringVal) {
+      int32_t err_msg_length;
+      TF_RETURN_IF_ERROR(client_->ReadInt(&err_msg_length));
+
+      uint8_t *err_msg_c = new uint8_t[err_msg_length];
+      auto clean = gtl::MakeCleanup([err_msg_c] { delete[] err_msg_c; });
```

Comment:
It might be OK though I am wondering if it is possible to convert
```
uint8_t *err_msg_c = new uint8_t[err_msg_length];
```

with smart pointers? `auto clean` releases the memory so not concern here.

But in the future if someone else touches this part of the code, having allocation and auto-release in one line with smart pointer, might be "safer" than having two lines of explicit allocate then auto-release.

Reply:
Yes, it's a good point. I updated the code, now it uses smart pointers.